### ISSUE TITLE
@Timed: histogram buckets have a too high cadinality

### ISSUE THREAD

**Issue by adericbourg** [micrometer-metrics/micrometer#1947]
Using `@Timed` annotation does not allow to control buckets for a timer (like the `sla()` method would on a timer builder). 

This results in having many buckets. 

* Most of them are not relevant
* It causes performance issues with TSBDs (at least Prometheus) by having a too high cardinality

Example (with a Prometheus registry): 

```
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.001",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.001048576",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.001398101",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.001747626",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.002097151",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.002446676",} 0.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="0.002796201",} 0.0

// 54 other lines here

http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="12.884901886",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="14.316557651",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="15.748213416",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="17.179869184",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="22.906492245",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="28.633115306",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="30.0",} 3.0
http_server_requests_seconds_bucket{exception="None",method="GET",outcome="SUCCESS",status="200",uri="/foo",le="+Inf",} 3.0
```

As a library user: 

* I'd like to benefit from the current integration
* I'd like to be able to define my own buckets

Currently, using buckets or not is enabled (if I read well) by the `histogram` attribute of `@Timed` annotation. 

I don't see a "good" option right now: 

* adding a new property (eg, an `sla` property) could bring confusion as it would conflict with `histogram`
* replacing `histogram` by `sla` would break the API

What do you think?

**Comment by jkschneider** [comment]
@adericbourg You can control the buckets added by `@Timed` with a `MeterFilter` that implements the `configure` method. Some shortcuts:

* `MeterFilter.maxExpected("http.server.requests", Duration.ofSeconds(1))`
* `MeterFilter.minExpected("http.server.reqests", Duration.ofMillis(10))`

`MeterFilter` implementations simply need to be wired as a `@Bean` in a Spring app to take effect.

Also, there is no need to use `@Timed` on Spring MVC and WebFlux endpoints, as these are automatically instrumented by the framework.

**Comment by adericbourg** [comment]
Thanks a lot for that answer: it is very helpful! 

I agree that tuning histogram may cause "bad" approximations and I don't mean to fight against it. 
Setting explicit bounds is part of the solution as the main issue I have here is that there are too many bounds. To provide some context, iour SRE team is struggling maintaining Prometheus instances that scrape metrics with a too high cardinality. My goal here is to set constraints on that cardinality.

A complete solution could be: 
* setting min and max bounds
* _and_ setting the number of bounds (number of buckets)

Anyhow, setting SLA manually and disabling `histogram` property is a good workaround for now.

**Comment by jkschneider** [comment]
> and setting the number of bounds (number of buckets)

This temptation is actually precisely why we removed (obvious) configurability. It's just too easy to say "hey it's publishing 100 buckets right now, can we turn that down to 50 buckets?" But the impact on percentile approximation is unknowable.

**Comment by fzyzcjy** [comment]
@jkschneider Hi thanks for the explanation - now I see why spring does not allow me to set the buckets explicitly, but only allow to set min/max. However, I see people recommending that, prometheus should not have too high cardinality, e.g. only 10 buckets should be used. What do you think about it, and how do you solve it in your production environment? Thanks

**Comment by renannprado** [comment]
TL;DR

```yaml
management:
  metrics:
    distribution:
      slo:
        http.server.requests:
          - 100ms
          - 500ms
          - 1s
          - 2s
```

**Comment by fjsabino022** [comment]
Hi guys, I am still facing a high cardinality of bucket numbers. How could you fix it?

**Comment by jonatan-ivanov** [comment]
I'm not sure why you enable percentiles and histogram too (you can calculate percentiles from the histogram on the backend so percentiles are redundant there).

The solution you can use right now is right above your comment: you can define SLOs one-by-one which will turn into histogram buckets.

**Comment by shakuzen** [comment]
In case anyone skipped over the reason why you probably don't actually want to hack your own explicit buckets using SLO configuration, see this from Jon's comment above:

> In general we try to discourage fine grained control of histogram buckets because experience has shown that folks tend to select buckets leading to high-error-bound percentile approximations (and the true error bound is in fact unknowable from the discretized distribution, so they never really know).
